src/diagram/models/dq-root.ts

```diff
@@ -1,31 +1,70 @@
-import { Instance, types, destroy } from "mobx-state-tree";
+import { Instance, types } from "mobx-state-tree";
 import { Elements, FlowTransform } from "react-flow-renderer/nocss";
-import { DQNode } from "./dq-node";
+import { DQNode, DQNodeType } from "./dq-node";
+import { VariableType } from "./variable";
+
+export interface VariablesAPI {
+  createVariable: () => VariableType;
+  removeVariable: (variable?: VariableType) => void;
+}
 
 export const DQRoot = types.model("DQRoot", {
-    nodes: types.map(DQNode),
-    flowTransform: types.maybe(types.frozen<FlowTransform>())
+  nodes: types.map(DQNode),
+  flowTransform: types.maybe(types.frozen<FlowTransform>())
 })
+.volatile(self => ({
+  variablesAPI: undefined as VariablesAPI | undefined
+}))
+.views(self => ({
+  get reactFlowElements() {
+    const elements: Elements = [];
+    self.nodes.forEach((node) => {
+      elements.push(...node.reactFlowElements);
+    });
+    return elements;
+  },
+  get nodeFromVariableMap() {
+    const map: Record<string, DQNodeType> = {};
+    self.nodes.forEach((node) => {
+      map[node.variableId] = node;
+    });
+    return map;
+  }
+}))
 .views(self => ({
-    get reactFlowElements() {
-        const elements: Elements = [];
-        self.nodes.forEach((node) => {
-            elements.push(...node.reactFlowElements);
-        });
-        return elements;
+  getNodeFromVariableId(id: string) {
+    return self.nodeFromVariableMap[id];
+  }
+}))
+.actions(self => ({
+  addNode(newNode: Instance<typeof DQNode>) {
```

Comment:
```suggestion
  addNode(newNode: DQNodeType) {
```
For consistency and readability